sympy/logic/algorithms/z3_wrapper.py

```
@@ -29,54 +75,139 @@ def z3_satisfiable(expr, all_models=False):
+                    uf_table[name] = max(uf_table[name] or 0, arity or 0)
+                else:
+                    uf_table[name] = max(uf_table[name], arity)
+            else:
+                uf_table[name] = arity
+    return uf_table
+
+
+def collect_all_free_symbols(enc_cnf):
+    """Collect all free symbols from all predicates in the CNF."""
+    free_symbols = set()
+    for pred, enc in enc_cnf.encoding.items():
+        if not is_supported_predicate(pred):
             continue
+        free_symbols |= pred.free_symbols
+    return free_symbols
 
-        pred_str = smtlib_code(pred, auto_declare=False, auto_assert=False, known_functions=known_functions)
 
-        symbols |= pred.free_symbols
-        pred = pred_str
-        clause = f"(implies d{enc} {pred})"
-        assertion = "(assert " + clause + ")"
-        assertions.append(assertion)
+def encoded_cnf_to_z3_solver(enc_cnf, z3):
+    s = z3.Solver()
+
+    uf_table = collect_all_uninterpreted_functions(enc_cnf)
+    free_symbols = collect_all_free_symbols(enc_cnf)
+
+    declarations = []
+    assertions = []
 
-    for sym in symbols:
+    declarations.extend(f"(declare-const d{var} Bool)" for var in enc_cnf.variables)
+    assertions.extend(clause_to_assertion(clause) for clause in enc_cnf.data)
+
+    # Create declarations for all uninterpreted functions
+    for name, arity in sorted(uf_table.items()):
+        if arity is None:
+            continue
+        elif arity == 0:
+            declarations.append(f"(declare-const {name} Real)")
+        else:
+            arg_types = " ".join(["Real"] * arity)
+            declarations.append(f"(declare-fun {name} ({arg_types}) Real)")
+
+    # Declare all free symbols that appear in the CNF
+    for sym in sorted(free_symbols, key=str):
         declarations.append(f"(declare-const {sym} Real)")
 
+    for pred, enc in enc_cnf.encoding.items():
+        if not is_supported_predicate(pred):
+            continue
```

If any predicates are not supported we should probably raise an exception